pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/KeyValueSchemaCompatibilityCheck.java

```diff
@@ -0,0 +1,95 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.broker.service.schema;
+
+import org.apache.pulsar.common.schema.KeyValue;
+import org.apache.pulsar.common.schema.SchemaData;
+import org.apache.pulsar.common.schema.SchemaType;
+
+import java.nio.ByteBuffer;
+
+/**
+ * {@link KeyValueSchemaCompatibilityCheck} for {@link SchemaType#KEY_VALUE}.
+ */
+public class KeyValueSchemaCompatibilityCheck extends AvroSchemaBasedCompatibilityCheck {
+
+    private KeyValue<byte[], byte[]> decode(byte[] bytes, SchemaData schemaData) {
+        ByteBuffer byteBuffer = ByteBuffer.wrap(bytes);
+        int keyLength = byteBuffer.getInt();
+        byte[] keySchema = new byte[keyLength];
+        byteBuffer.get(keySchema);
+
+        int valueLength = byteBuffer.getInt();
+        byte[] valueSchema = new byte[valueLength];
+        byteBuffer.get(valueSchema);
+        return new KeyValue<>(keySchema, valueSchema);
+    }
+
+    @Override
+    public SchemaType getSchemaType() {
+        return SchemaType.KEY_VALUE;
+    }
+
+    @Override
+    public boolean isCompatible(SchemaData from, SchemaData to, SchemaCompatibilityStrategy strategy) {
+        KeyValue<byte[], byte[]> fromKeyValue = this.decode(from.getData(), from);
+        KeyValue<byte[], byte[]> toKeyValue = this.decode(to.getData(), to);
+
+        SchemaData fromKeySchemaData = SchemaData.builder().data(fromKeyValue.getKey())
+                .type(SchemaType.valueOf(from.getProps().get("key.schema.type"))).build();
+
+        SchemaData fromValueSchemaData = SchemaData.builder().data(fromKeyValue.getValue())
+                .type(SchemaType.valueOf(from.getProps().get("value.schema.type"))).build();
+
+        SchemaData toKeySchemaData = SchemaData.builder().data(toKeyValue.getKey())
+                .type(SchemaType.valueOf(to.getProps().get("key.schema.type"))).build();
+
+
+        SchemaData toValueSchemaData = SchemaData.builder().data(toKeyValue.getValue())
+                .type(SchemaType.valueOf(to.getProps().get("value.schema.type"))).build();
+
+        JsonSchemaCompatibilityCheck jsonSchemaCompatibilityCheck = new JsonSchemaCompatibilityCheck();
+        if (SchemaType.valueOf(to.getProps().get("key.schema.type")) == SchemaType.AVRO
```

Comment:
You don't need the following logic here. The whole logic in JsonSchemaCompatibilityCheck was there for backward compatibility check. A key type and a value type can be any schema types. It shouldn't be limited to JSON or 
AVRO as well.

You should pass `Map<SchemaType, SchemaCompatibilityCheck> checkers` in to KeyValueSchemaCompatibilityCheck in its constructor.

```
private final Map<SchemaType, SchemaCompatibilityCheck> checkers;

public KeyValueSchemaCompatibilityCheck(Map<SchemaType, SchemaCompatibilityCheck> checkers) {
    this.checkers = checkers;
}
```

Then the logic is much simpler.

```
SchemaType fromKeyType = SchemaType.valueOf(from.getProps().get("key.schema.type"));
SchemaType fromValueType = SchemaType.valueOf(from.getProps().get("value.schema.type"));
SchemaType toKeyType = SchemaType.valueOf(to.getProps().get("key.schema.type"));
SchemaType toValueType = SchemaType.valueOf(to.getProps().get("value.schema.type"));

if (fromKeyType != toKeyType || fromValueType != toValueType) {
    return false;
}

KeyValue<byte[], byte[]> fromKeyValueSchemaData = this.splitKeyValueSchemaData(from.getData());
KeyValue<byte[], byte[]> toKeyValueSchemaData = this.splitKeyValueSchemaData(to.getData());

SchemaData fromKeySchemaData = ...;
SchemaData fromValueSchemaData = ...;
SchemaData toKeySchemaData = ...;
SchemaData toValueSchemaData = ...;

SchemaCompatibilityCheck keyCheck = checks.get(keySchemaType);
SchemaCompatibilityCheck valueCheck = checks.get(valueSchemaType);

return keyCheck.isCompatible(fromKeySchemaData, toKeySchemaData, strategy) && valueCheck.isCompatible(fromValueSchemaData, toValueSchemaData, strategy);
```